### PR TITLE
[10804 ] Collection of data and notification to listeners DISCOVERED_ENTITY

### PR DESCRIPTION
**Pull request by MiguelBarro**
Merge after https://github.com/eProsima/Fast-DDS/pull/1919

### CODE REVIEW

File: test/unittest/statistics/rtps/RTPSStatisticsTests.cpp [hunk 1]
```diff
@@ -41,181 +47,709 @@ namespace rtps {
...
+
+        int loop = 0;
+        while ( callbacks < 5 )
+        {
+            this_thread::sleep_for(chrono::milliseconds(100));
+            if ( ++loop > 10 )
+            {
+                break;
+            }
+        }
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/statistics/rtps/RTPSStatisticsTests.cpp:747]: I am missing a check that the expected callbacks are correctly received. Also, maybe this waiting can be substituted with a block until the expected callbacks are received.

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 2]
```diff
@@ -698,6 +701,9 @@ ReaderProxyData* PDP::addReaderProxyData(
     logInfo(RTPS_PDP, "Adding reader proxy data " << reader_guid);
     ReaderProxyData* ret_val = nullptr;
 
+    // notify statistics module
+    getRTPSParticipant()->on_entity_discovery(reader_guid);
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp:705]: I don't think so. The use case is to check if the discovery data is received, disregarding if it is managed properly.

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 3]
```diff
@@ -790,6 +796,9 @@ WriterProxyData* PDP::addWriterProxyData(
     logInfo(RTPS_PDP, "Adding reader proxy data " << writer_guid);
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp:796]: that's an old one 🤣

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 4]
```diff
@@ -790,6 +796,9 @@ WriterProxyData* PDP::addWriterProxyData(
...
     WriterProxyData* ret_val = nullptr;
 
+    // notify statistics module
+    getRTPSParticipant()->on_entity_discovery(writer_guid);
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp:800]: I don't think so. The use case is to check if the discovery data is received, disregarding if it is managed properly.

File: test/unittest/statistics/rtps/RTPSStatisticsTests.cpp (hunk 1 above)
Comment (MiguelBarro) [reply test/unittest/statistics/rtps/RTPSStatisticsTests.cpp:759]: There is a `EXPECT_CALL` above that checks if the callbacks are received.
The only purpose of the loop is to provide some time for the remote entities to be created and PDP and EDP endpoints to communicate. If we introduce a condition variable then the test will timeout if the callbacks are not performed, that is not what we want.

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Fix linters (it probably comes from a previous PR)

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM

**Pull request merged**